### PR TITLE
Rick and Morty - Step 4

### PR DESCRIPTION
**Pull request by LetiRocha**
Added a Recycler view with its adapter to show all the characters received from backend in a list 
When data is obtained from network, we hid the button cause it's not relevant anymore

## Summary by CodeRabbit

- **New Features**
  - Introduced a new `CharacterListAdapter` to manage and display a dynamic list of characters in a RecyclerView.
  - Added a new layout for character details, featuring an image, name, and status within a CardView.

- **Enhancements**
  - Updated `MainActivity` to support dynamic character list updates from a remote API.
  - Improved UI by displaying a list of characters and removing redundant UI elements.

- **Bug Fixes**
  - Fixed issues related to data updates and UI refreshes upon retrieving new character data.

### CODE REVIEW

File: app/src/main/res/layout/activity_main.xml [hunk 1]
```diff
@@ -7,47 +7,24 @@
...
-        android:layout_marginBottom="16dp"
-        android:text="@string/first_button"
-        app:layout_constraintBottom_toTopOf="@id/character_name"
-        app:layout_constraintEnd_toEndOf="parent"
-        app:layout_constraintStart_toStartOf="parent" />
-
-    <TextView
-        android:id="@+id/character_name"
+    <androidx.recyclerview.widget.RecyclerView
+        android:id="@+id/character_list"
```
[the remark this answers is not in the document]
Comment (LetiRocha) [reply app/src/main/res/layout/activity_main.xml:11]: Fixed

File: app/src/main/res/layout/activity_main.xml [hunk 2]
```diff
@@ -7,47 +7,24 @@
...
-    <ImageView
-        android:id="@+id/character_image"
-        android:layout_width="0dp"
-        android:layout_height="0dp"
-        app:layout_constraintDimensionRatio="1:1"
-        app:layout_constraintEnd_toEndOf="parent"
-        app:layout_constraintStart_toStartOf="parent"
-        app:layout_constraintTop_toBottomOf="@id/character_status"
-        app:layout_constraintWidth_percent="0.2" />
+        app:layout_constraintTop_toBottomOf="@id/character_list" />
```
[the remark this answers is not in the document]
Comment (LetiRocha) [reply app/src/main/res/layout/activity_main.xml]: Fixed

File: app/src/main/res/layout/activity_main.xml [hunk 3]
```diff
@@ -7,47 +7,24 @@
...
-        app:layout_constraintBottom_toTopOf="@id/character_name"
-        app:layout_constraintEnd_toEndOf="parent"
-        app:layout_constraintStart_toStartOf="parent" />
-
-    <TextView
-        android:id="@+id/character_name"
-        android:layout_width="wrap_content"
+    <androidx.recyclerview.widget.RecyclerView
+        android:id="@+id/character_list"
+        android:layout_width="match_parent"
```
[the remark this answers is not in the document]
Comment (LetiRocha) [reply app/src/main/res/layout/activity_main.xml:12]: No I don't want to change that, this is intented. And it's the second time you tell me the same

File: app/src/main/java/com/mstudio/superheroarch/CharacterListAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        val character = characterList[position]
+        holder.characterNameTextView.text = character.name
+        holder.characterStatusTextView.text = character.status
+        Picasso.get().load(character.image).into(holder.characterImageView);
+    }
+
+    fun updateData(newItems: List<Character>) {
+        characterList.clear()
+        characterList.addAll(newItems)
+        notifyDataSetChanged()
```
Comment (PabloGrelaM) [inline app/src/main/java/com/mstudio/superheroarch/CharacterListAdapter.kt:38]: it's ok, but this line is deprecated, we'll need a fix later.

File: app/src/main/java/com/mstudio/superheroarch/MainActivity.kt [hunk 5]
```diff
@@ -31,28 +33,35 @@ class MainActivity : AppCompatActivity() {
...
             insets
         }
-        val button = findViewById<Button>(R.id.button_to_change_text)
-        characterNameTitle = findViewById(R.id.character_name)
-        characterStatusTitle = findViewById(R.id.character_status)
+        updateButton = findViewById(R.id.refresh_button)
+
+        val characterListRecyclerView: RecyclerView = findViewById(R.id.character_list)
+        characterListRecyclerView.layoutManager = LinearLayoutManager(this)
+        characterListAdapter = CharacterListAdapter(mutableListOf())
```
Comment (PabloGrelaM) [inline app/src/main/java/com/mstudio/superheroarch/MainActivity.kt:40]: you don't need to add the list in the constructor as you are not using it here.

File: app/src/main/java/com/mstudio/superheroarch/MainActivity.kt [hunk 6]
```diff
@@ -31,28 +33,35 @@ class MainActivity : AppCompatActivity() {
...
             override fun onResponse(call: Call<CharactersResponse>, response: Response<CharactersResponse>) {
                 if (response.isSuccessful) {
-                    val firstCharacter = response.body()?.results?.firstOrNull()
-                    firstCharacter?.let {
-                        characterNameTitle?.text = getString(R.string.character_name, it.name)
-                        characterStatusTitle?.text = getString(R.string.character_status, it.status)
-                        Picasso.get().load(it.image).into(findViewById<ImageView>(R.id.character_image));
-                    } ?: Snackbar.make(findViewById(R.id.main), getString(R.string.error_message_no_character_info), Snackbar.LENGTH_SHORT).show()
+                    response.body()?.results?.let {
+                        for (character in response.body()?.results ?: emptyList()) {
```
Comment (PabloGrelaM) [inline app/src/main/java/com/mstudio/superheroarch/MainActivity.kt:58]: you don't need this for. Maybe is a mistake due to the naming. The val "result" is characterList. With that name you'll see easily that you just need to do response.body()?.characterList?.let { listOfCharacters.add(it) }

**Review by PabloGrelaM — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/mstudio/superheroarch/MainActivity.kt [hunk 7]
```diff
@@ -1,26 +1,27 @@
...
-import com.squareup.picasso.Picasso
 import retrofit2.Call
 import retrofit2.Callback
 import retrofit2.Response
 
 
 class MainActivity : AppCompatActivity() {
-    var characterNameTitle: TextView? = null
-    var characterStatusTitle: TextView? = null
+    lateinit var characterListAdapter: CharacterListAdapter
```
Comment (PabloGrelaM) [inline app/src/main/java/com/mstudio/superheroarch/MainActivity.kt:22]: you don't need this lateinit now

**Review by PabloGrelaM — APPROVED** [review]

**Review by YagoAccenture — APPROVED** [review]

**Pull request merged**